### ISSUE TITLE
Crash when trying to export all

### ISSUE THREAD

**Issue by maxolasersquad** [OpenTracksApp/OpenTracks#491]
**Describe the bug**
When I try to export all of my tracks the application crashes.

**To Reproduce**
1. Tap the ellipses in the top-right
2. Tap settings
3. Tap "Export all"
4. Browse to the desired destination and then tap the "Use this folder" button
5. The application crashes.

**Technical information**
 - Pixel 3 XL
 - Android 11
 - OpenTracks v3.11.1-a42bd432e installed from F-Droid.

**Comment by dennisguse** [comment]
Any chance to get a logcat?

**Comment by rgmf** [comment]
I've had the oportunity of reproduce it in a Pixel 3a with Android 11 and last F-Droid OT's version. Here is what logcat says:

```
java.lang.SecurityException: Permission Denial: writing com.android.externalstorage.ExternalStorageProvider uri content://com.android.externalstorage.documents/tree/primary%3AOpenTracksBackup/document/primary%3AOpenTracksBackup from pid=26717, uid=10293 requires android.permission.MANAGE_DOCUMENTS, or grantUriPermission()
        at android.content.ContentProvider.enforceWritePermissionInner(ContentProvider.java:896)
        at com.android.externalstorage.ExternalStorageProvider.enforceWritePermissionInner(ExternalStorageProvider.java:154)
        at android.provider.DocumentsProvider.callUnchecked(DocumentsProvider.java:1119)
        at android.provider.DocumentsProvider.call(DocumentsProvider.java:1067)
        at com.android.externalstorage.ExternalStorageProvider.call(ExternalStorageProvider.java:647)
        at android.content.ContentProvider.call(ContentProvider.java:2448)
        at android.content.ContentProvider$Transport.call(ContentProvider.java:517)
        at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:295)
        at android.os.Binder.execTransactInternal(Binder.java:1154)
        at android.os.Binder.execTransact(Binder.java:1123)
2020-11-01 17:46:39.032 26717-26963/? E/AndroidRuntime: FATAL EXCEPTION: Thread-2
    Process: de.dennisguse.opentracks, PID: 26717
    java.lang.NullPointerException: Attempt to invoke virtual method 'android.net.Uri androidx.documentfile.provider.DocumentFile.getUri()' on a null object reference
        at de.dennisguse.opentracks.fragments.ExportProgressDialogFragment$ExportThread.exportTrack(ExportProgressDialogFragment.java:219)
        at de.dennisguse.opentracks.fragments.ExportProgressDialogFragment$ExportThread.run(ExportProgressDialogFragment.java:200)
```

PS/ I also have the issue that was described in https://github.com/OpenTracksApp/OpenTracks/issues/439:

> Another minor issue is, when exporting a second time, the folder is shown, but the button at the bottom "use this folder for Export" is missing frowning_face

**Comment by maxolasersquad** [comment]
I had to do some research on how to generate logcat logs.

```
11-01 12:17:22.728  3834  3900 E DatabaseUtils: Writing exception to parcel
11-01 12:17:22.728  3834  3900 E DatabaseUtils: java.lang.SecurityException: Permission Denial: reading com.android.externalstorage.ExternalStorageProvider uri content://com.android.externalstorage.documents/tree/primary%3AOpenTracks/document/primary%3AOpenTracks/children from pid=31012, uid=10457 requires that you obtain access using ACTION_OPEN_DOCUMENT or related APIs
11-01 12:17:22.728  3834  3900 E DatabaseUtils: 	at android.content.ContentProvider.enforceReadPermissionInner(ContentProvider.java:820)
11-01 12:17:22.728  3834  3900 E DatabaseUtils: 	at com.android.externalstorage.ExternalStorageProvider.enforceReadPermissionInner(ExternalStorageProvider.java:147)
11-01 12:17:22.728  3834  3900 E DatabaseUtils: 	at android.content.ContentProvider$Transport.enforceReadPermission(ContentProvider.java:684)
11-01 12:17:22.728  3834  3900 E DatabaseUtils: 	at android.content.ContentProvider$Transport.query(ContentProvider.java:239)
11-01 12:17:22.728  3834  3900 E DatabaseUtils: 	at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:106)
11-01 12:17:22.728  3834  3900 E DatabaseUtils: 	at android.os.Binder.execTransactInternal(Binder.java:1154)
11-01 12:17:22.728  3834  3900 E DatabaseUtils: 	at android.os.Binder.execTransact(Binder.java:1123)
11-01 12:17:22.729 31012  4008 W DocumentFile: Failed query: java.lang.SecurityException: Permission Denial: reading com.android.externalstorage.ExternalStorageProvider uri content://com.android.externalstorage.documents/tree/primary%3AOpenTracks/document/primary%3AOpenTracks/children from pid=31012, uid=10457 requires that you obtain access using ACTION_OPEN_DOCUMENT or related APIs
11-01 12:17:22.730  3834  3900 E DatabaseUtils: Writing exception to parcel
11-01 12:17:22.730  3834  3900 E DatabaseUtils: java.lang.SecurityException: Permission Denial: writing com.android.externalstorage.ExternalStorageProvider uri content://com.android.externalstorage.documents/tree/primary%3AOpenTracks/document/primary%3AOpenTracks from pid=31012, uid=10457 requires android.permission.MANAGE_DOCUMENTS, or grantUriPermission()
11-01 12:17:22.730  3834  3900 E DatabaseUtils: 	at android.content.ContentProvider.enforceWritePermissionInner(ContentProvider.java:896)
11-01 12:17:22.730  3834  3900 E DatabaseUtils: 	at com.android.externalstorage.ExternalStorageProvider.enforceWritePermissionInner(ExternalStorageProvider.java:154)
11-01 12:17:22.730  3834  3900 E DatabaseUtils: 	at android.provider.DocumentsProvider.callUnchecked(DocumentsProvider.java:1119)
11-01 12:17:22.730  3834  3900 E DatabaseUtils: 	at android.provider.DocumentsProvider.call(DocumentsProvider.java:1067)
11-01 12:17:22.730  3834  3900 E DatabaseUtils: 	at com.android.externalstorage.ExternalStorageProvider.call(ExternalStorageProvider.java:647)
11-01 12:17:22.730  3834  3900 E DatabaseUtils: 	at android.content.ContentProvider.call(ContentProvider.java:2448)
11-01 12:17:22.730  3834  3900 E DatabaseUtils: 	at android.content.ContentProvider$Transport.call(ContentProvider.java:517)
11-01 12:17:22.730  3834  3900 E DatabaseUtils: 	at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:295)
11-01 12:17:22.730  3834  3900 E DatabaseUtils: 	at android.os.Binder.execTransactInternal(Binder.java:1154)
11-01 12:17:22.730  3834  3900 E DatabaseUtils: 	at android.os.Binder.execTransact(Binder.java:1123)
11-01 12:17:22.731 31012  4008 W DocumentsContract: Failed to create document
11-01 12:17:22.731 31012  4008 W DocumentsContract: java.lang.SecurityException: Permission Denial: writing com.android.externalstorage.ExternalStorageProvider uri content://com.android.externalstorage.documents/tree/primary%3AOpenTracks/document/primary%3AOpenTracks from pid=31012, uid=10457 requires android.permission.MANAGE_DOCUMENTS, or grantUriPermission()
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at android.os.Parcel.createExceptionOrNull(Parcel.java:2373)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at android.os.Parcel.createException(Parcel.java:2357)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at android.os.Parcel.readException(Parcel.java:2340)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:190)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:142)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at android.content.ContentProviderProxy.call(ContentProviderNative.java:732)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at android.content.ContentResolver.call(ContentResolver.java:2405)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at android.provider.DocumentsContract.createDocument(DocumentsContract.java:1371)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:53)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:45)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at de.dennisguse.opentracks.io.file.ExportFile.getExportDocumentFile(ExportFile.java:13)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at de.dennisguse.opentracks.fragments.ExportProgressDialogFragment$ExportThread.exportTrack(ExportProgressDialogFragment.java:212)
11-01 12:17:22.731 31012  4008 W DocumentsContract: 	at de.dennisguse.opentracks.fragments.ExportProgressDialogFragment$ExportThread.run(ExportProgressDialogFragment.java:200)
11-01 12:17:22.732 31012  4008 E AndroidRuntime: FATAL EXCEPTION: Thread-2
11-01 12:17:22.732 31012  4008 E AndroidRuntime: Process: de.dennisguse.opentracks, PID: 31012
11-01 12:17:22.732 31012  4008 E AndroidRuntime: java.lang.NullPointerException: Attempt to invoke virtual method 'android.net.Uri androidx.documentfile.provider.DocumentFile.getUri()' on a null object reference
11-01 12:17:22.732 31012  4008 E AndroidRuntime: 	at de.dennisguse.opentracks.fragments.ExportProgressDialogFragment$ExportThread.exportTrack(ExportProgressDialogFragment.java:219)
11-01 12:17:22.732 31012  4008 E AndroidRuntime: 	at de.dennisguse.opentracks.fragments.ExportProgressDialogFragment$ExportThread.run(ExportProgressDialogFragment.java:200)
11-01 12:17:22.742  3329 15701 I MediaProvider: Begin Intent { act=android.intent.action.MEDIA_SCANNER_SCAN_FILE dat=file:///storage/emulated/0/OpenTracks/1.gpx flg=0x1000010 cmp=com.google.android.providers.media.module/com.android.providers.media.MediaService }
11-01 12:17:22.743  1465  4013 I DropBoxManagerService: add tag=data_app_crash isTagEnabled=true flags=0x2
11-01 12:17:22.744  1465  3473 W ActivityTaskManager:   Force finishing activity de.dennisguse.opentracks/.io.file.exporter.ExportActivity
11-01 12:17:22.759 31012  4008 I Process : Sending signal. PID: 31012 SIG: 9
11-01 12:17:22.760  1465  1530 W BroadcastQueue: Background execution not allowed: receiving Intent { act=android.intent.action.DROPBOX_ENTRY_ADDED flg=0x10 (has extras) } to com.google.android.gms/.stats.service.DropBoxEntryAddedReceiver
11-01 12:17:22.761  1465  1530 W BroadcastQueue: Background execution not allowed: receiving Intent { act=android.intent.action.DROPBOX_ENTRY_ADDED flg=0x10 (has extras) } to com.google.android.gms/.chimera.GmsIntentOperationService$PersistentTrustedReceiver
11-01 12:17:22.773  3684  3709 I AdrenoGLES-0: QUALCOMM build                   : c9bef8a, Ie73904e3bd
11-01 12:17:22.773  3684  3709 I AdrenoGLES-0: Build Date                       : 06/26/20
11-01 12:17:22.773  3684  3709 I AdrenoGLES-0: OpenGL ES Shader Compiler Version: EV031.31.04.00
11-01 12:17:22.773  3684  3709 I AdrenoGLES-0: Local Branch                     : gfx-adreno.lnx.2.0
11-01 12:17:22.773  3684  3709 I AdrenoGLES-0: Remote Branch                    : quic/gfx-adreno.lnx.2.0
11-01 12:17:22.773  3684  3709 I AdrenoGLES-0: Remote Branch                    : NONE
11-01 12:17:22.773  3684  3709 I AdrenoGLES-0: Reconstruct Branch               : NOTHING
11-01 12:17:22.773  3684  3709 I AdrenoGLES-0: Build Config                     : C P 11.0.1 AArch64
11-01 12:17:22.773  3684  3709 I AdrenoGLES-0: Driver Path                      : /vendor/lib64/egl/libGLESv2_adreno.so
11-01 12:17:22.773   926   926 I Zygote  : Process 31012 exited due to signal 9 (Killed)
11-01 12:17:22.778  1465  1478 I ActivityManager: Process de.dennisguse.opentracks (pid 31012) has died: fg  TOP 
11-01 12:17:22.778  1465  1905 W InputDispatcher: channel '248218 de.dennisguse.opentracks/de.dennisguse.opentracks.settings.SettingsActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
11-01 12:17:22.778  1465  1905 E InputDispatcher: channel '248218 de.dennisguse.opentracks/de.dennisguse.opentracks.settings.SettingsActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
11-01 12:17:22.778  1465  2193 I WindowManager: WIN DEATH: Window{710c054 u0 Please wait…}
11-01 12:17:22.778  1465  2193 W InputDispatcher: Attempted to unregister already unregistered input channel '710c054 Please wait… (server)'
11-01 12:17:22.779  1465  1538 I libprocessgroup: Successfully killed process cgroup uid 10457 pid 31012 in 1ms
11-01 12:17:22.780  3329 15701 I MediaProvider: End Intent { act=android.intent.action.MEDIA_SCANNER_SCAN_FILE dat=file:///storage/emulated/0/OpenTracks/1.gpx flg=0x1000010 cmp=com.google.android.providers.media.module/com.android.providers.media.MediaService }
11-01 12:17:22.783  1465  2114 I WindowManager: WIN DEATH: Window{248218 u0 de.dennisguse.opentracks/de.dennisguse.opentracks.settings.SettingsActivity}
11-01 12:17:22.783  1465  2114 W InputDispatcher: Attempted to unregister already unregistered input channel '248218 de.dennisguse.opentracks/de.dennisguse.opentracks.settings.SettingsActivity (server)'
11-01 12:17:22.783  3684  3709 I AdrenoGLES-0: PFP: 0x016ee189, ME: 0x00000000
11-01 12:17:22.784  1465  3991 I WindowManager: WIN DEATH: Window{dc0609f u0 de.dennisguse.opentracks/de.dennisguse.opentracks.io.file.exporter.ExportActivity}
11-01 12:17:22.784  1465  3991 W InputDispatcher: Attempted to unregister already unregistered input channel 'dc0609f de.dennisguse.opentracks/de.dennisguse.opentracks.io.file.exporter.ExportActivity (server)'
11-01 12:17:22.786  1465  1478 W ActivityTaskManager: Force removing ActivityRecord{962118e u0 de.dennisguse.opentracks/.settings.SettingsActivity t58866}: app died, no saved state
11-01 12:17:22.793  1465  1520 D CompatibilityChangeReporter: Compat change id reported: 135634846; UID 10457; state: DISABLED
11-01 12:17:22.793  1465  1531 D CompatibilityChangeReporter: Compat change id reported: 143937733; UID 10457; state: ENABLED
11-01 12:17:22.816   926   926 D Zygote  : Forked child process 4019
11-01 12:17:22.818  1465  1531 I ActivityManager: Start proc 4019:de.dennisguse.opentracks/u0a457 for top-activity {de.dennisguse.opentracks/de.dennisguse.opentracks.TrackListActivity}
11-01 12:17:22.825  4019  4019 E guse.opentrack: Not starting debugger since process cannot load the jdwp agent.
11-01 12:17:22.863  4019  4019 D NetworkSecurityConfig: No Network Security Config specified, using platform default
11-01 12:17:22.863  4019  4019 D NetworkSecurityConfig: No Network Security Config specified, using platform default
11-01 12:17:22.864  2652  2157 I NetworkScheduler.Stats: Task com.google.android.gms/com.google.android.gms.ipa.base.IpaGcmTaskService started execution. cause:9 exec_start_elapsed_seconds: 4495 [CONTEXT service_id=218 ]
11-01 12:17:22.867  1465  1521 W WindowManager: Failed to deliver inset state change to w=Window{248218 u0 de.dennisguse.opentracks/de.dennisguse.opentracks.settings.SettingsActivity EXITING}
11-01 12:17:22.867  1465  1521 W WindowManager: android.os.DeadObjectException
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.os.BinderProxy.transactNative(Native Method)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.os.BinderProxy.transact(BinderProxy.java:540)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.view.IWindow$Stub$Proxy.insetsControlChanged(IWindow.java:765)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at com.android.server.wm.WindowState.notifyInsetsControlChanged(WindowState.java:3669)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at com.android.server.wm.InsetsStateController.lambda$notifyPendingInsetsControlChanged$4$InsetsStateController(InsetsStateController.java:445)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at com.android.server.wm.-$$Lambda$InsetsStateController$-1iOXDf-1s3wDHcMIHBKNk6MS3I.run(Unknown Source:2)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at com.android.server.wm.WindowAnimator.executeAfterPrepareSurfacesRunnables(WindowAnimator.java:341)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at com.android.server.wm.RootWindowContainer.performSurfacePlacementNoTrace(RootWindowContainer.java:854)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at com.android.server.wm.RootWindowContainer.performSurfacePlacement(RootWindowContainer.java:802)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at com.android.server.wm.WindowSurfacePlacer.performSurfacePlacementLoop(WindowSurfacePlacer.java:178)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at com.android.server.wm.WindowSurfacePlacer.performSurfacePlacement(WindowSurfacePlacer.java:127)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at com.android.server.wm.WindowManagerService.relayoutWindow(WindowManagerService.java:2279)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at com.android.server.wm.Session.relayout(Session.java:213)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.view.ViewRootImpl.relayoutWindow(ViewRootImpl.java:7443)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2635)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1952)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:8171)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:972)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.view.Choreographer.doCallbacks(Choreographer.java:796)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.view.Choreographer.doFrame(Choreographer.java:731)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:957)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.os.Handler.handleCallback(Handler.java:938)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.os.Handler.dispatchMessage(Handler.java:99)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.os.Looper.loop(Looper.java:223)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at android.os.HandlerThread.run(HandlerThread.java:67)
11-01 12:17:22.867  1465  1521 W WindowManager: 	at com.android.server.ServiceThread.run(ServiceThread.java:44)
```

**Comment by maxolasersquad** [comment]
This does appear to be a duplicate of https://github.com/OpenTracksApp/OpenTracks/issues/439 which is marked as Closed (which I guess is why I didn't find it when searching).

**Comment by maxolasersquad** [comment]
It looks like #439 was attempted to be fixed in https://github.com/OpenTracksApp/OpenTracks/commit/bc04efbaac52772f13764b250b4d3957d909dd25 which was just before 3.11.0 was released, and I'm on 3.11.1. Am I seeing a slightly different bug, or maybe the fix in that commit didn't fully resolve it?

**Comment by dennisguse** [comment]
Looks like we miss a permission.
Can try to add android.permission.MANAGE_DOCUMENTS to the manifest and see if this solves the issue?

**Comment by rgmf** [comment]
Afte reading the new changes in Android 11, I think we need doing this: https://developer.android.com/preview/privacy/storage#all-files-access

More info: https://medium.com/androiddevelopers/android-11-storage-faq-78cefea52b7c

PS/ With MANAGE_DOCUMENTS in the manifest doesn't work for me.

**Comment by pstorch** [comment]
I think I found something. Will send a PR for testing.

**Comment by maxolasersquad** [comment]
@pstorch If you can get me an apk I'll be happy to install it and see if it works.

**Comment by pstorch** [comment]
@maxolasersquad this is a debug build with a possible fix:
[OpenTracks-debug.zip](https://github.com/OpenTracksApp/OpenTracks/files/5477015/OpenTracks-debug.zip)

Can be installed in parallel to the release version.

**Comment by dennisguse** [comment]
I would love to move this on into the upcoming release - we are slightly overdue ;)
Any news on this one?

**Comment by pstorch** [comment]
The fix worked for @rgmf and for me in the emulator. I don't know if @maxolasersquad had time to do a third test on his physical device.

**Comment by maxolasersquad** [comment]
I finally got to test it this evening. It worked perfectly.
